File: examples/notify.js
#!/usr/bin/env node
/**
 * NotificationHub – Node.js Client-Klasse + Beispiele
 *
 * Usage:
 *   NHUB_URL=http://localhost:3000 NHUB_TOKEN=your-token node examples/notify.js
 */

const NHUB_URL   = process.env.NHUB_URL   ?? 'http://localhost:3000';
const NHUB_TOKEN = process.env.NHUB_TOKEN ?? 'changeme';

class NotificationHub {
  constructor(baseUrl = NHUB_URL, token = NHUB_TOKEN) {
    this.baseUrl = baseUrl;
    this.token   = token;
  }

  #headers() {
    return {
      'Content-Type': 'application/json',
      'Authorization': `Bearer ${this.token}`,
    };
  }

  async login(email, password) {
    const res = await fetch(`${this.baseUrl}/api/v1/auth/login`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ email, password }),
    });
    if (!res.ok) throw new Error(await res.text());
    const data = await res.json();
    this.token = data.token;
    return data.user;
  }

  async notify({ source, title, message, priority = 'INFO', ...rest }) {
    const res = await fetch(`${this.baseUrl}/api/v1/notify`, {
      method: 'POST',
      headers: this.#headers(),
      body: JSON.stringify({ source, title, message, priority, ...rest }),
    });
    if (!res.ok) throw new Error(await res.text());
    return res.json();
  }

  async getNotifications({ page = 1, limit = 20, ...filters } = {}) {
    const params = new URLSearchParams({ page, limit, ...filters });
    const res = await fetch(`${this.baseUrl}/api/v1/notifications?${params}`, {
      headers: this.#headers(),
    });
    if (!res.ok) throw new Error(await res.text());
    return res.json();
  }

  async getStats() {
    const res = await fetch(`${this.baseUrl}/api/v1/notifications/stats/summary`, {
      headers: this.#headers(),
    });
    if (!res.ok) throw new Error(await res.text());
    return res.json();
  }

  async acknowledge(id) {
    const res = await fetch(`${this.baseUrl}/api/v1/notifications/${id}/acknowledge`, {
      method: 'POST', headers: this.#headers(),
    });
    if (!res.ok) throw new Error(await res.text());
    return res.json();
  }
}

// ---- Beispiele ---------------------------------------------------------------
const hub = new NotificationHub();

(async () => {
  console.log('--- Proxmox: Node offline ---');
  const r1 = await hub.notify({
    source:   'Proxmox',
    service:  'pve-manager',
    title:    'Node pve03 nicht erreichbar',
    message:  'pve03 antwortet nicht mehr auf Cluster-Heartbeats. Letzter Kontakt vor 2 Minuten.',
    priority: 'EMERGENCY',
    hostname: 'pve03',
    tags:     ['cluster', 'node', 'offline'],
  });
  console.log(r1);

  console.log('\n--- Zertifikat läuft ab ---');
  const r2 = await hub.notify({
    source:   'CertBot',
    title:    'TLS-Zertifikat läuft in 7 Tagen ab',
    message:  'Das Zertifikat für example.com läuft am 2026-08-05 ab. Bitte erneuern.',
    priority: 'WARNING',
    hostname: 'proxy01',
    tags:     ['ssl', 'certificate'],
    extra:    { domain: 'example.com', expires: '2026-08-05' },
  });
  console.log(r2);

  console.log('\n--- Stats ---');
  const stats = await hub.getStats();
  console.log(`Heute: ${stats.today} | Critical: ${stats.critical} | Warnings: ${stats.warnings}`);
})().catch(console.error);

File: examples/notify.py
#!/usr/bin/env python3
"""
NotificationHub – Python-Hilfsbibliothek + Beispiele
Install: pip install requests

Usage:
  NHUB_URL=http://localhost:3000 NHUB_TOKEN=your-token python3 examples/notify.py
"""
import os
import requests
from dataclasses import dataclass, field
from typing import Any, Optional

NHUB_URL   = os.getenv("NHUB_URL",   "http://localhost:3000")
NHUB_TOKEN = os.getenv("NHUB_TOKEN", "changeme")


@dataclass
class NotificationHubClient:
    base_url: str = NHUB_URL
    token: str    = NHUB_TOKEN
    timeout: int  = 5

    def _headers(self) -> dict:
        return {
            "Content-Type": "application/json",
            "Authorization": f"Bearer {self.token}",
        }

    def login(self, email: str, password: str) -> None:
        """Token via Login holen und im Client setzen."""
        resp = requests.post(
            f"{self.base_url}/api/v1/auth/login",
            json={"email": email, "password": password},
            timeout=self.timeout,
        )
        resp.raise_for_status()
        self.token = resp.json()["token"]

    def notify(
        self,
        source: str,
        title: str,
        message: str,
        priority: str = "INFO",
        service: Optional[str] = None,
        hostname: Optional[str] = None,
        tags: list[str] = field(default_factory=list),
        extra: dict[str, Any] = field(default_factory=dict),
        **kwargs,
    ) -> dict:
        payload = {
            "source":   source,
            "title":    title,
            "message":  message,
            "priority": priority,
            "tags":     tags or [],
            "extra":    extra or {},
        }
        if service:  payload["service"]  = service
        if hostname: payload["hostname"] = hostname
        payload.update(kwargs)

        resp = requests.post(
            f"{self.base_url}/api/v1/notify",
            headers=self._headers(),
            json=payload,
            timeout=self.timeout,
        )
        resp.raise_for_status()
        return resp.json()

    def get_notifications(self, page: int = 1, limit: int = 20, **filters) -> dict:
        params = {"page": page, "limit": limit, **filters}
        resp = requests.get(
            f"{self.base_url}/api/v1/notifications",
            headers=self._headers(),
            params=params,
            timeout=self.timeout,
        )
        resp.raise_for_status()
        return resp.json()

    def get_stats(self) -> dict:
        resp = requests.get(
            f"{self.base_url}/api/v1/notifications/stats/summary",
            headers=self._headers(),
            timeout=self.timeout,
        )
        resp.raise_for_status()
        return resp.json()

    def acknowledge(self, notification_id: str) -> dict:
        resp = requests.post(
            f"{self.base_url}/api/v1/notifications/{notification_id}/acknowledge",
            headers=self._headers(),
            timeout=self.timeout,
        )
        resp.raise_for_status()
        return resp.json()


# ---- Beispiele --------------------------------------------------------------
if __name__ == "__main__":
    import json

    hub = NotificationHubClient()

    # Optional: Token via Login holen statt direkt übergeben
    # hub.login("admin@localhost", "changeme")

    print("--- Proxmox Backup OK ---")
    result = hub.notify(
        source="Proxmox",
        service="vzdump",
        title="Backup vm-100 erfolgreich",
        message="Backup von vm-100 fertig. Dauer: 4m 12s, Größe: 8.3 GB.",
        priority="SUCCESS",
        hostname="pve01",
        tags=["backup", "proxmox"],
    )
    print(json.dumps(result, indent=2))

    print("\n--- Disk kritisch ---")
    result = hub.notify(
        source="NodeExporter",
        service="filesystem",
        title="Disk /data auf nas01 kritisch",
        message="/data ist zu 97% belegt. Noch 58 GB frei.",
        priority="CRITICAL",
        hostname="nas01",
        tags=["disk", "storage"],
        extra={"percent": 97, "free_gb": 58},
    )
    print(json.dumps(result, indent=2))

    print("\n--- Stats ---")
    stats = hub.get_stats()
    print(f"Heute: {stats['today']} | Critical: {stats['critical']} | Warnings: {stats['warnings']}")
